domain/bankaccount.py

```diff
@@ -1,43 +1,68 @@
 from typing import List
-from domain.transaction import Transaction, create_transaction
-from domain.commontypes.types import AccountID
 
-Amount = float
+from domain.ledgeraccount import LedgerAccount
+from domain.transaction import Transaction, create_transaction
+from domain.commontypes.types import AccountID, Amount
 
 
 class BankAccount:
     def __init__(self,
                  account_id: AccountID,
-                 transactions: List[Transaction]
+                 main_account: LedgerAccount,
+                 draft_account: LedgerAccount,
+                 stage_account: LedgerAccount,
+                 overdraft_limit: Amount
                  ):
         self._id = account_id
-        self._transactions = transactions
+        self._main_account = main_account
+        self._draft_account = draft_account
+        self._stage_account = stage_account
+        self._overdraft_limit = overdraft_limit
 
     def get_balance(self) -> Amount:
-        balance = 0.0
-        for t in self._transactions:
-            if t.d_acc == self._id:
-                balance -= t.value
-            else:
-                balance += t.value
-        return balance
+        main_balance = self._main_account.get_balance()
+        draft_balance = self._draft_account.get_balance()
+        return main_balance - draft_balance
 
     def transfer(self, destiny_id: AccountID, value: Amount) -> None:
         if value <= 0:
             raise InvalidValueToTransfer(value)
 
-        balance = self.get_balance()
-        if balance < value:
-            raise InsufficientFundsException(balance, value)
+        main_balance = self._main_account.get_balance()
+        main_id = self._main_account.account_id
+        if main_balance < value:
+            remaining = value - main_balance
+            draft_account_balance = self._draft_account.get_balance()
+            overdraft = self._overdraft_limit - draft_account_balance
+
+            if overdraft >= remaining:
+                draft_id = self._draft_account.account_id
+                draft_transaction = create_transaction(draft_id, main_id, remaining)
+                main_transaction = create_transaction(main_id, destiny_id, value)
+
+                self._draft_account.add_transaction(draft_transaction)
+                self._main_account.add_transaction(draft_transaction)
+                self._main_account.add_transaction(main_transaction)
+
+            else:
+                raise InsufficientFundsException(self.get_balance(), value)
         else:
-            transaction = create_transaction(self._id, destiny_id, value)
-            self._transactions.insert(0, transaction)
+
+            transaction = create_transaction(main_id, destiny_id, value)
```

Comment:
There's no transaction between a ledgerAccount (main_id) and a bankAccount (destiny_id). All transfer to a bank account must be to their stage's ledgerAccount. This only works because create_transaction is wrongly typed, using int, and not ID's. That shows that we must have two types of transactions : ledgerTransactions and bankTransactions, the first between ledgeraccounts and the other between bankAccounts.

Reply:
So what should I do:
1 - Use the given destiny_id to find the stage's id from destination bankAccount
OR
2 - The destiny_id should now be treated as a destiny_stage_id

or another option i could not think about?

Reply:
Using the first option, I can get a Maybe[BankAccount] using accounts repo (but the get_by_id is wrong because of our recent changes).

Reply:
To type the create_transaction's arguments, should I use native id, UUID or create my own stuff?

Reply:
"So what should I do:
1 - Use the given destiny_id to find the stage's id from destination bankAccount
OR
2 - The destiny_id should now be treated as a destiny_stage_id"

1

"
Using the first option, I can get a Maybe[BankAccount] using accounts repo (but the get_by_id is wrong because of our recent changes).
"

you don't need use the repository. The conversor between bank account and ledger account that lies on that bank account must live in the domain, with a pure function. I sugest to use UUID to bank account, and a string with format "UUID-xxx" where xxx is a 3 digit number. The main of a bank account with id foo is foo-1, draft foo-2 and stage foo-3.

Don't think about repository now, we will change on next comments.

"
To type the create_transaction's arguments, should I use native id, UUID or create my own stuff?
"

You must have two types of transactions, LedgerTransaction and BankAccountTransaction, and each must have their own create_transaction, with correct types.

The types must be abstract, AccountId and LedgerId , but the concrete types can be UUIDv4 and a dataclass with UUIDv4 and 3-digit suffix.

Reply:
"you don't need use the repository. The conversor between bank account and ledger account that lies on that bank account must live in the domain, with a pure function. " ─ That means I will need a accounts_conversor on domain level and this conversor will receive a accountID and do its magic to create a BankAccount then get its LedgerAccount or receive a BankAccount and only return its stage_account?

Also, how do we construct a LedgerAccount to then construct a BankAccount?

Reply:
Something like to_tranfer(bankId : AccountId) : LedgerId. You don't need to load an object of BankAccount, is just a map between id's. And yes, live in domain.